lib/task/task.rb

```diff
@@ -0,0 +1,144 @@
+require 'virtus'
+require 'active_support/concern'
+require_relative 'data_interface/interface'
+
+module Task
+  # A Task represents a task to be performed. The task can be serialized as JSON to be passed to
+  # other processes to complete, and can be stored in Cassandra as a means of ensuring "at-least-once"
+  # task completion for this task. So, an example lifecycle of a task would be:
+  #   1) A task is generated in a "master" job
+  #   2) The master job saves the task, providing a record that the task was generated.
+  #   3) The master job passes the task to another job to complete the task.
+  #   4) The worker job completes task, removing the record for that task.
+  #   5) If the worker job fails, the task is not completed and so the record for it persists.
+  #   6) A subsequent job can fetch the list of tasks, returning the tasks that failed to complete.
+  #      This job could either serialize them to be completed by other jobs, or complete them directly.
+  #
+  # A task has a unique id, belongs to a task list that group similar tasks together, and has data
+  # associated with it.
+  #
+  # @example Defining a type of Task
+  #   class MyDeleteTask
+  #     include Task::Task
+  #     # This task has the item_id data field, representing the item to delete
+  #     data_attr_reader :item_id
+  #   end
+  #
+  # @example Creating a task
+  #   my_task = MyDeleteTask.build(:id => 'my_id', :task_list => "#{service.id}-delete", :item_id => '123')
+  #
+  # @example Serializing and deserializing a task
+  # # In the process that generated the task
+  # serialized = my_task.as_hash
+  #
+  # # In the process that acts on the hash
+  # my_task_copy = Task::Task.from_hash(serialized)
+  #
+  # @example Saving a task
+  # my_task.save
+  #
+  # @example Fetching a single task that has been saved
+  # Task::DataInterface::Interface.new.find(task_list, task_id)
+  #
+  # @example Fetching all tasks for a task list
+  # Task::DataInterface::Interface.new.all(task_list)
+  #
+  # @example Completing a task, so that it is not longer fetchable
+  # my_task.complete
+  #
+  module Task
+    extend ActiveSupport::Concern
+    include Virtus.module
+
+    attribute :task_list, String
+    attribute :id, String
+    attribute :data, Hash[Symbol => Object]
+
+    module ClassMethods
+      # Instantiate an instance of this Task subclass.
+      # @param options [Hash] Options to instantiate this Task. :task_list and :id are required;
+      #   other arguments will be passed as data to the task.
+      # @option options [String] :task_list
+      # @option options [String] :id
+      def build(options)
```

Comment:
I don't think it's really a factory pattern, it's just an alternative to `new` that automatically places attributes in the data hash. I didn't want to override the default (Virtus) initializer.